extensions/livedata/src/main/java/com/twocoders/extensions/livedata/LiveDataExtensions.kt

```diff
@@ -0,0 +1,382 @@
+        if (firstEmitted && secondEmitted) {
+            value = combineFunction(firstValue as A, secondValue as B)
+        }
+    }
+
+    addSource(second) {
+        secondEmitted = true
+        secondValue = it
+        if (firstEmitted && secondEmitted) {
+            value = combineFunction(firstValue as A, secondValue as B)
+        }
+    }
+}
+
+/**
+ * @see combineLatest
+ */
+@MainThread
+fun <A, B, C> combineLatest(first: LiveData<A>, second: LiveData<B>, third: LiveData<C>) =
+    combineLatest(first, second, third) { a, b, c -> Triple(a, b, c) }
+
+/**
+ * @see combineLatest
+ */
+@MainThread
+@Suppress("UNCHECKED_CAST")
+fun <A, B, C, R> combineLatest(
+    first: LiveData<A>,
+    second: LiveData<B>,
+    third: LiveData<C>,
+    combineFunction: (A, B, C) -> R
+) = MediatorLiveData<R>().apply {
+
+    var firstEmitted = false
+    var firstValue: A? = null
+
+    var secondEmitted = false
+    var secondValue: B? = null
+
+    var thirdEmitted = false
+    var thirdValue: C? = null
+
+    addSource(first) {
+        firstEmitted = true
+        firstValue = it
+        if (firstEmitted && secondEmitted && thirdEmitted) {
+            value = combineFunction(firstValue as A, secondValue as B, thirdValue as C)
+        }
+    }
+
+    addSource(second) {
+        secondEmitted = true
+        secondValue = it
+        if (firstEmitted && secondEmitted && thirdEmitted) {
+            value = combineFunction(firstValue as A, secondValue as B, thirdValue as C)
+        }
+    }
+
+    addSource(third) {
+        thirdEmitted = true
+        thirdValue = it
+        if (firstEmitted && secondEmitted && thirdEmitted) {
+            value = combineFunction(firstValue as A, secondValue as B, thirdValue as C)
+        }
+    }
+}
+
+/**
+ * zips both of the LiveData and emits a value after both of them have emitted their values,
+ * after that, waits again for both LiveData to emit new value.
+ *
+ * The difference between combineLatest and zip is that the zip only emits after all LiveData
+ * objects have a new value, but combineLatest will emit after any of them has a new value.
+ */
+@MainThread
+fun <A, B> LiveData<A>.zip(second: LiveData<B>): LiveData<Pair<A, B>> = zip(this, second)
+
+@MainThread
+fun <A, B> zip(first: LiveData<A>, second: LiveData<B>) = zip(first, second) { a, b -> Pair(a, b) }
+
+@MainThread
+@Suppress("UNCHECKED_CAST")
+fun <A, B, R> zip(first: LiveData<A>, second: LiveData<B>, zipFunction: (A, B) -> R) = MediatorLiveData<R>().apply {
+
+    var firstEmitted = false
+    var firstValue: A? = null
+
+    var secondEmitted = false
+    var secondValue: B? = null
+
+    addSource(first) {
+        firstEmitted = true
+        firstValue = it
+        if (firstEmitted && secondEmitted) {
+            value = zipFunction(firstValue as A, secondValue as B)
+            firstEmitted = false
+            secondEmitted = false
+        }
+    }
+
+    addSource(second) {
+        secondEmitted = true
```

Comment:
v zasade to nemusi vadit, ale rx zip tusim neprepisuje hodnoty tak ako tato funkcia. napr.

`A1, B1, A2, A3, B2` tu vrati -> `A1B1, A3B3`; v rx by to poslalo `A1B1, A2B2` a `A3` by cakalo na dalsie B

Reply:
Nepsal jsem tuto extension a nevím jak je to přesně v Rx, ale já bych v tomto tvém příkladě právě čekal, že přijde fresh z obou dvou, tedy A2 už mě nezajímá, když mezi tím přišlo A3 než dorazilo B2.

Reply:
nj, to len taka poznamka na okraj... ono by to malo fungovat ako "zips" (teda vzdy sa caka 1 k 1 v rx)